RFD 1: use testify/require instead of testify/assert

awly opened the pull request:
`testify/require` has the same API as `testify/assert`, but it aborts the test on failure (which we almost always want).
Updated all existing tests to match.

Also fix examples in the RFD

Commit:
Fix example code in RFD 1

All `assert.X` functions take `testing.T` as first argument.
Also, add example of `assert.ErrorAssertionFunc` in subtests.

kimlisa commented on rfd/0001-testing-guidelines.md:
```diff
@@ -117,21 +117,21 @@ Subtests
 			got, gotErr := parseInt(tt.in)
-			assert.Equal(err != nil, tt.wantErr, "got error:", err)
-			assert.Equal(got, tt.want)
+            tt.assertErr(t, err)
```
spacing, it looks like this:
![image](https://user-images.githubusercontent.com/43280172/95919234-137f5480-0d62-11eb-8402-ec04f9309923.png)

kimlisa replied:
also, should it be `tt.assertErr(t, gotErr)`? or maybe `gotErr` to `err`

awly replied:
fixed

kimlisa commented on rfd/0001-testing-guidelines.md:
```diff
@@ -117,21 +117,21 @@ Subtests
-		{desc: "empty", in: "", wantErr: true},
+		{desc: "positive", in: "123", want: 123, assertErr: assert.NoError},
+		{desc: "negative", in: "-123", want: -123, assertErr: assert.NoError},
+		{desc: "non-numeric", in: "abc", assertErr: assert.Error},
+		{desc: "empty", in: "", assertErr: assert.Error},
```
these are missing `want` field, for the `assert.Equal(t, got, tt.want)` check

kimlisa replied:
or does the `assert.Error` skips this check?

awly replied:
Nice catch. I just realized that we should be using `testify/require` instead of `testify/assert`.
They have the same API but `require` terminates the test on failure, while `assert` lets it proceed.

With `require`, error is checked first and if that check fails, `want` is never checked.

Commit:
RFD 1: user testify/require instead of testify/assert

`require` is a sister package to `assert` that terminates the test on
failure. `assert` records the failure but lets the test proceed, which
is un-intuitive.

Also update all existing tests to match.

awly reviewed (commented):
Thanks @kimlisa, PTAL

fspmarshall reviewed (approved):
If possible, we should no longer vendor `testify/assert`.

awly commented:
@fspmarshall `testify/require` depends on `testify/assert` (it's a wrapper essentially).

need approval from @klizhentas or @russjones for `/vendor` changes

kimlisa commented on rfd/0001-testing-guidelines.md:
```diff
@@ -117,21 +117,21 @@ Subtests
-		{desc: "empty", in: "", wantErr: true},
+		{desc: "positive", in: "123", want: 123, assertErr: require.NoError},
+		{desc: "negative", in: "-123", want: -123, assertErr: require.NoError},
+		{desc: "non-numeric", in: "abc", assertErr: require.Error},
+		{desc: "empty", in: "", assertErr: require.Error},
```
> With require, error is checked first and if that check fails, want is never checked.

Sorry andrew, i'm still confused.

`assertErr: require.Error`, this means we are asserting that it returned an error (not nil), so this is a check that passes. Doesn't that mean the `require.Equal(t, got, tt.want)` will still run? (meaning we still need want values)

awly replied:
ah, right, it still runs but we assume that the function under test returns the zero value of `want` (0)

kimlisa replied:
ah i get it now! i forgot about the zero values if not explicitly defined.

awly commented:
This still (somehow) needs approval from @benarent or @webvictim

benarent reviewed (approved):
I'm going to get more people to approve .md files.

webvictim commented:
![image](https://user-images.githubusercontent.com/897821/96199931-0cf20800-0f2f-11eb-806a-2f5554e8190a.png)
